Adding recipe for api-validation filter.

james49 opened the pull request.

james49 commented:
This is an atom hopper based configuration.  And, as such, needs Saxon licensing.

b0d0nne11 commented on recipes/default.rb:
```diff
@@ -43,7 +43,8 @@
   'translation'           => node['repose']['translation'          ]['cluster_id'],
-  'uri-identity'          => node['repose']['uri_identity'         ]['cluster_id']
+  'uri-identity'          => node['repose']['uri_identity'         ]['cluster_id'],
+  'api-validator'         => node['repose']['api_validator'        ]['cluster_id']
```
Super nit pick, but the rest of these are alphabetical. Would you mind moving this up the list?

james49 replied:
I don't mind making it alph-beta.  I was attempting to defer to the 'standard filter order' suggested by the wiki, which has some loose recommendations:  https://repose.atlassian.net/wiki/display/REPOSE/Filters+and+services .   But if that's not an issue, then I'll adjust it.

james49 replied:
Made the change in latest commit.

b0d0nne11 commented on templates/default/validator.cfg.xml.erb:
```diff
@@ -0,0 +1,66 @@
+	<validator
+        role="<%= validator['role'] %>"
+
+        <% unless validator['enable_api_coverage'].empty? %>
```
Would `validator.has_key? 'enable_api_coverage'` be more descriptive maybe?

james49 replied:
Made the change for all optional elements.

b0d0nne11 commented on recipes/filter-api-validator.rb:
```diff
@@ -0,0 +1,35 @@
+  notifies :restart, 'service[repose-valve]'
+end
+
+cookbook_file "allfeeds_observer.wadl" do
```
These files probably need to be configurable since I imagine different products will want to bring their own wadls. We could do this by making the `cookbook` attribute of these `cookbook_file` resources configurable. Something like this maybe?

```ruby
node['repose']['api_validator']['wadl_filenames'].each do |f|
  cookbook_file f do
    cookbook node['repose']['api_validator']['wadl_cookbook']
    path "#{node['repose']['config_directory']}/#{f}.wadl"
    action :create_if_missing
  end
end
```

james49 replied:
Excellent idea!  Thanks for the code snippet.

james49 replied:
Made the change in latest commit.  Since a wadl is associated with each validator element, the recipe iterates through each validator in the configuration attributes and renders the file name from there.

b0d0nne11 commented on recipes/filter-api-validator.rb:
```diff
@@ -0,0 +1,35 @@
+node['repose']['api_validator']['validators'].each do |f|
+  if f.has_key?("wadl") 
+    wfile = f['wadl']
+    cookbook_file wfile do
```
I think we still need to expose the cookbook attribute here so that ppl can bring there own without having to extend the cookbook.

james49 replied:
Don't understand what 'cookbook attribute' you are talking about exposing in your last comment.  Can you clarify?

b0d0nne11 replied:
There is a cookbook attribute on the cookbook_file resource that allows you to use a file in another cookbook.

http://docs.chef.io/resource_cookbook_file.html#attributes

If we set it to `node['repose']['api_validator']['wadl_cookbook']` and make the default `repose` it will allow others to bring there own wadl files in another cookbook without having to modify this one.

b0d0nne11 commented on attributes/default.rb:
```diff
@@ -192,3 +192,37 @@
+default['repose']['api_validator']['cluster_id'] = ['all']
+default['repose']['api_validator']['multi_role_match'] = true 
+default['repose']['api_validator']['delegating_quality'] = 0.6 
+default['repose']['api_validator']['validators'] = [
```
Looks like the defaults are specific to cloud feeds and require a lincense. Can we make them a little more generic? Feeds should be able to set up their validators in whatever role or wrapper cookbook they're using.

b0d0nne11 commented:
Hey @james49, thanks a lot for the update. I like that you included the wadl filename in the validator configuration hash. That makes a lot of sense. I added a few notes. I also noticed that there are a couple of rubocop warnings regarding trailing whitespace and tabs.

james49 commented:
I removed the AH specific details and made the configuration more generic.  Also, fixed the rubocop warnings that were showing up (specific to the api-validator part).

filler commented:
@james49 whats the status of this PR?  Do you still want to move forward on it?  If so, youll need to rebase.

Let me know.  Thanks!

dimtruck commented:
@filler I think we should abandon this.  I haven't seen anything more being done with it and I _believe_ we are out of date with v7 anyway.

filler commented:
![](https://media.giphy.com/media/SZioIIBxB7QRy/giphy.gif)
